File: app/src/main/java/com/doubean/ford/ui/groups/groupDetail/GroupDetailViewModel.java
package com.doubean.ford.ui.groups.groupDetail;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

import com.doubean.ford.data.repository.GroupRepository;
import com.doubean.ford.data.repository.GroupsFollowsAndSavesRepository;
import com.doubean.ford.data.vo.GroupDetail;
import com.doubean.ford.data.vo.Resource;

public class GroupDetailViewModel extends ViewModel {
    private final GroupRepository groupRepository;
    private final LiveData<Resource<GroupDetail>> group;
    private final GroupsFollowsAndSavesRepository groupsFollowsAndSavesRepository;
    private final String groupId;
    private final LiveData<Boolean> isGroupFollowed;

    public GroupDetailViewModel(GroupRepository groupRepository, GroupsFollowsAndSavesRepository groupsFollowsAndSavesRepository, String groupId, String defaultTabId) {
        this.groupsFollowsAndSavesRepository = groupsFollowsAndSavesRepository;
        this.groupId = groupId;
        this.groupRepository = groupRepository;
        group = groupRepository.getGroup(groupId, true);
        this.isGroupFollowed = groupsFollowsAndSavesRepository.getFollowed(groupId, null);
    }

    public LiveData<Boolean> getFollowed() {
        return isGroupFollowed;
    }

    public LiveData<Resource<GroupDetail>> getGroup() {
        return group;
    }

    public void addFollow() {
        groupsFollowsAndSavesRepository.createFollow(groupId, null);
    }

    public void removeFollow() {
        groupsFollowsAndSavesRepository.removeFollow(groupId, null);
    }
}

File: app/src/main/java/com/doubean/ford/ui/groups/groupDetail/GroupDetailViewModel.kt
package com.doubean.ford.ui.groups.groupDetail

import androidx.lifecycle.*
import com.doubean.ford.data.prefs.DataStorePreferenceStorage
import com.doubean.ford.data.repository.GroupRepository
import com.doubean.ford.data.repository.GroupUserDataRepository
import com.doubean.ford.model.PostSortBy
import com.doubean.ford.util.Event
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class GroupDetailViewModel(
    groupRepository: GroupRepository,
    private val groupUserDataRepository: GroupUserDataRepository,
    private val preferenceStorage: DataStorePreferenceStorage,
    private val groupId: String,
) : ViewModel() {

    private val _pagerPreselectedEvent = MutableLiveData(Event(Unit))
    val pagerPreselectedEvent: LiveData<Event<Unit>> = _pagerPreselectedEvent

    val group = groupRepository.getGroup(groupId).flowOn(Dispatchers.IO).asLiveData()

    fun addFollow() {
        viewModelScope.launch {
            groupUserDataRepository.addFollowedGroup(
                groupId = groupId,
                enablePostNotifications = preferenceStorage.perFollowDefaultEnablePostNotifications.first(),
                allowsDuplicateNotifications = preferenceStorage.perFollowDefaultAllowDuplicateNotifications.first(),
                sortRecommendedPostsBy = preferenceStorage.perFollowDefaultSortRecommendedPostsBy.first(),
                feedRequestPostCountLimit = preferenceStorage.perFollowDefaultFeedRequestPostCountLimit.first()
            )
        }

    }

    fun removeFollow() {
        viewModelScope.launch {
            groupUserDataRepository.removeFollowedGroup(groupId)
        }
    }

    fun saveNotificationsPreference(
        enableNotifications: Boolean,
        allowNotificationUpdates: Boolean,
        sortRecommendedPostsBy: PostSortBy,
        numberOfPostsLimitEachFeedFetch: Int,
    ) {
        viewModelScope.launch {
            withContext(Dispatchers.IO) {
                groupUserDataRepository.updateGroupNotificationsPref(
                    groupId,
                    enableNotifications,
                    allowNotificationUpdates,
                    sortRecommendedPostsBy,
                    numberOfPostsLimitEachFeedFetch
                )
            }

        }
    }

    companion object {
        class Factory(
            private val repository: GroupRepository,
            private val groupUserDataRepository: GroupUserDataRepository,
            private val preferenceStorage: DataStorePreferenceStorage,
            private val groupId: String,
            private val defaultTab: String?,
        ) : ViewModelProvider.NewInstanceFactory() {
            @Suppress("UNCHECKED_CAST")
            override fun <T : ViewModel> create(modelClass: Class<T>): T {
                return GroupDetailViewModel(
                    repository,
                    groupUserDataRepository,
                    preferenceStorage,
                    groupId
                ) as T
            }

        }
    }
}